make an initial README for how to contribute a new SPIFFE project (#263)
spiffe/spiffe: NEW_PROJECTS.md

> # Creating a new SPIFFE Project

Reviewer: Suggest adding an `## Introduction` section.

a brief section at the beginning that summarizes what SPIFFE is and why someone might be interested in contributing. This would help set the context for the reader.

Author: added

Reviewer 2: This doc references the creation of new SPIFFE projects, but you may also consider addressing project _donations_ as well. Perhaps we don't see a difference between those things?

We are grateful for your desire to expand the SPIFFE ecosystem with your contribution!

This is an evolving set of guidelines and instructions for how to bring your existing project, or idea for a completely fresh project, under the SPIFFE organization.

If your project is already under the SPIFFE banner, you may be looking for [MATURITY](/MATURITY.md) instead.

## Notice

Maintainer groups are given significant freedom in managing their projects to that we may best foster growth and innovation. There are some elements of SPIFFE subproject ownership to be aware of:

> The [SPIFFE Steering Committee](/ssc/README.md) (SSC) governs the SPIFFE organization as a whole, and has the final authority on acceptance, rejection, or eviction of projects. Additionally, as stewards of community health, the SSC is responsible for intervening in any [Code of Conduct](/CODE-OF-CONDUCT.md) violations or derelictions of maintainer responsibility.

Reviewer: These responsibilities would be easier to follow (on a web-based medium) if they were bullet points.

Author: updated

Note that when contributing a project to SPIFFE, its purpose becomes to serve the SPIFFE community rather than any personal gain. Additionally, we expect a sufficient number of maintainers for a project as it grows, not all of whom may bring the same direction you may originally have had in mind.
Carefully consider this change in control and purpose.

*Every SPIFFE project* at minimum must follow the [SPIFFE Code of Conduct](/CODE-OF-CONDUCT.md).

## Project Proposal

### Contact

The SSC can be contacted with proposal details at ssc@spiffe.io or on [our SPIFFE slack channel](https://spiffe.slack.com/archives/C01GQ267JJU).

### Proposal Outline

The following details are expected for proposals to help us best understand how the project will serve the community.

Additionally, the information here make great initial `README.md`, `CONTRIBUTING.md`, and `GOVERNANCE.md` files for the repo!

#### Initial Maintainer List

The list of initial maintainers of the project, including GitHub handle.

> #### Project Purpose and Direction

Reviewer: we can consider providing example questions that can guide contributors in framing their project’s objectives and vision.

Author: suggestions to add beyond what's here?
I also want to provide as much guide here as possible; current list is all I could initially think of but happy to add more

Reviewer: For purpose

- What specific pain point does your project address?
- What are the current workarounds or alternative solutions, and why are they inadequate?
- Who are the primary end-users or beneficiaries of this project?
- How does your project align with SPIFFE's overall mission and goals?

For Direction
- What is the project’s roadmap for the next 6-12 months?
- How will the project scale? What are the scalability goals?
- Are there any legal or ethical considerations?
- What are potential challenges, and how do you plan to overcome them?
- What kind of community involvement do you foresee (e.g., are there opportunities for novice contributors, or is the project more suitable for experts in the field?)
- What technologies will the project leverage, and are they aligned with the larger ecosystem?
- How will the project maintain data integrity, security, and privacy?
- Are there any possible collaborations with existing SPIFFE projects?
- How will your project deal with failure scenarios or fallbacks?
- 
Providing these questions as a guide can help contributors delve deeper into their project's objectives and execution plan, offering a more comprehensive view for the review committee.

But on the flipside; adding too much of these questions can feel like a burden.
I’m not sure, maybe just a subset of the most important ones?

Author: added

What need is the project solving for the community? What problem exists or what enhancement will this create?

What is the *direction* of the project? This is similar but distinct from the *purpose* of the project.

While the purpose highlights the "Why" and "What" of the project, the direction should be about the "How".
This could include:
- What does "done" mean for the project? What is the idealized vision for the shape of the project? What are the milestones to complete for each level of [maturity](/MATURITY.md)?
- What are practices or styles that will be encouraged or discouraged?
- What notable features or options are in-scope or out-of-scope for the project? This especially will help differentiate your project from any similar ones that might already exist.

It's ok if there's not fully fleshed-out answers to each of these, and projects evolve over time. However the more information that can be provided the more easily SSC may understand the proposal, and the more confidence there will be in alignment amongst the maintainers.

> #### Governance

Reviewer: perhaps add a bit about conflict resolution mechanisms within the team of maintainers.

Author: under Governance or Purpose/Direction? I could see either being fitting

Reviewer: Not sure… “Governance” is better I think.

Author: added

The operations of the project.

How will maintainers be added or removed? What is the goal number of maintainers?

How often will the maintainers meet?

How will maintainers interact with contributors and review their Pull Requests and Issues? Slack? Dedicated online meetings? What is the goal turn around time?

#### Similar Projects

Do similar projects already exist actively within the SPIFFE community? If so, what makes this project sufficiently unique to justify any overlap?

Do similar projects exist outside of SPIFFE that can better help us understand the project's value?

Has this project existed in the past but was abandoned or closed out? What is different in today's world?

> ## Project Review

Reviewer: Shall we specify a timeframe for the review process?

Reviewer: For the review, would we require any documents? pre-meetings?

Reviewer: + 
Offering examples and template propsal(s) can streamline the process and clarify some doubts.

Reviewer: Maybe a FAQ section (though, the doc is satisfactory enough in itself)

Author: > Shall we specify a timeframe for the review process?

if within SSC we can come up with a consensus soon on a timeframe, sure. I was anticipating for this initial draft that that may not be something we could do right now, especially without real practice at it

> Offering examples and template propsal(s) can streamline the process and clarify some doubts.

We could take as a follow-up PR, or ask some existing projects to retroactively go through this process so we can point to theirs?

Author: added "a couple SSC sync cycles" as initial guidance. we really should be able to have a decision within that time period I think for most cases

If your project is already under the SPIFFE banner, you may be looking for [MATURITY](/MATURITY.md) instead.

## Notice

Maintainer groups are given significant freedom in managing their projects to that we may best foster growth and innovation. There are some elements of SPIFFE subproject ownership to be aware of:
- The [SPIFFE Steering Committee](/ssc/README.md) (SSC) governs the SPIFFE organization as a whole, and has the final authority on acceptance, rejection, or eviction of projects. Additionally, as stewards of community health, the SSC is responsible for intervening in any [Code of Conduct](/CODE-OF-CONDUCT.md) violations or derelictions of maintainer responsibility.
- Note that when contributing a project to SPIFFE, its purpose becomes to serve the SPIFFE community rather than any personal gain. Additionally, we expect a sufficient number of maintainers for a project as it grows, not all of whom may bring the same direction you may originally have had in mind.

Carefully consider this change in control and purpose.

*Every SPIFFE project* at minimum must follow the [SPIFFE Code of Conduct](/CODE-OF-CONDUCT.md).

## Project Proposal

### Contact

> The SSC can be contacted with proposal details at ssc@spiffe.io or on [our SPIFFE slack channel](https://spiffe.slack.com/archives/C01GQ267JJU).

Reviewer 2: I took a swing at changing this a little bit to include the slack signup link in case the reader isn't a member there yet ... just an idea

```suggestion
The SSC can be contacted with proposal details at ssc@spiffe.io or in the #ssc channel in the [SPIFFE slack](https://slack.spiffe.io). 
```

Author: > This doc references the creation of new SPIFFE projects, but you may also consider addressing project donations as well.

yup, in my head I was considering these the same thing. But good to be explicit :)
